### PR TITLE
{lyn13480} Make Asset Cache Server feature configurable at run time 

### PR DESCRIPTION
**Pull request by jackalbe**
## What does this PR do?

The Asset Cache Server (ACS) is presently only configurable at start time. 
This changes the logic to allow the ACS feature to be modified at run time.

## How was this PR tested?

Unit tests were updated to test new configuration logic.

### CODE REVIEW

File: Code/Tools/AssetProcessor/native/utilities/AssetServerHandler.cpp [hunk 1]
```diff
@@ -60,21 +116,120 @@ namespace AssetProcessor
...
+    {
+        AZStd::string previousServerAddress = m_serverAddress;
+        m_serverAddress = address;
+        if (!IsServerAddressValid())
+        {
+            m_serverAddress = previousServerAddress;
+            AZ_Error(AssetProcessor::DebugChannel, false,
+                "Server address (%s) is invalid! Reverting back to (%s)",
+                address.c_str(),
+                previousServerAddress.c_str());
```
Comment (AMZN-stankowi) [inline Code/Tools/AssetProcessor/native/utilities/AssetServerHandler.cpp:229]: you could use %.*s and AZ_STRING_ARG(previousServerAddress) for AZStd::string printouts. I don't know if you want to change it, it doesn't really make the code more or less readable.

File: Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h [hunk 2]
```diff
@@ -242,10 +249,31 @@ namespace AssetProcessor
         //! and put them in the temporary directory provided by the builderParam.
         //! This will return true if it was able to retrieve all the relevant job data from the server, otherwise return false.
         virtual bool RetrieveJobResult(const AssetProcessor::BuilderParams& builderParams) = 0;
+        //! Retrive the current mode for shared caching
```
Comment (AMZN-stankowi) [inline Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h:252]: retrive -> retrieve

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/resourcecompiler/RCBuilder.cpp [hunk 3]
```diff
@@ -288,23 +351,21 @@ namespace AssetProcessor
...
-            // intentionaly using a set here, as we want it to be the same order each time for hashing.
+            // intentionally using a set here, as we want it to be the same order each time for hashing.
             AZStd::set<AZStd::string> fingerprintRelevantParameters;
 
-            for (auto internalAssetRecognizer : *internalRecognizerList)
+            for (auto* internalAssetRecognizer : internalRecognizerList.second)
             {
                 // so referring to the structure explanation above, internalAssetRecognizer is
-                // one of those objects that has the regex in it, (along with list of commands to apply per platform)
+                // one of those objects that has the REGEX in it, (along with list of commands to apply per platform)
```
[the remark this answers is not in the document]
Comment (jackalbe) [reply Code/Tools/AssetProcessor/native/resourcecompiler/RCBuilder.cpp:368]: it was getting a squiggly line underneath it, maybe RegEx?

File: Code/Tools/AssetProcessor/native/utilities/AssetServerHandler.cpp (hunk 1 above)
Comment (jackalbe) [reply Code/Tools/AssetProcessor/native/utilities/AssetServerHandler.cpp:229]: all of the cool kids are doing it :)

File: Code/Tools/AssetProcessor/native/utilities/ApplicationManager.cpp [hunk 4]
```diff
@@ -488,6 +489,10 @@ bool ApplicationManager::StartAZFramework()
     //Registering all the Components
     m_frameworkApp.RegisterComponentDescriptor(AzFramework::LogComponent::CreateDescriptor());
 
+    AZ::SerializeContext* serializeContext{};
+    AZ::ComponentApplicationBus::BroadcastResult(serializeContext, &AZ::ComponentApplicationRequests::GetSerializeContext);
+    AssetProcessor::PlatformConfiguration::Reflect(serializeContext);
+
     Reflect();
```
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/utilities/ApplicationManager.cpp:496]: There's a Reflect call right below here, should this code go in there?

File: Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h [hunk 5]
```diff
@@ -242,10 +249,31 @@ namespace AssetProcessor
...
+        virtual void SetRemoteCachingMode(AssetServerMode mode) = 0;
+        //! Retrieve the remote folder location for the shared cache 
+        virtual const AZStd::string& GetServerAddress() const = 0;
+        //! Store the remote folder location for the shared cache 
+        virtual void SetServerAddress(const AZStd::string& address) = 0;
     };
-
     using AssetServerBus = AZ::EBus<AssetServerBusTraits>;
 
+    // This EBUS is notify listeners when Asset Server state(s) changes.
```
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h:263]: Typo: "EBUS is notify listeners"

File: Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h [hunk 6]
```diff
@@ -242,10 +249,31 @@ namespace AssetProcessor
...
+    class AssetServerNotifications
+        : public AZ::EBusTraits
+    {
+    public:
+        static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Multiple; // multi listener
+        static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single; //single bus
+        typedef AZStd::recursive_mutex MutexType;
+
+        //! TBD
```
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/utilities/AssetUtilEBusHelper.h:272]: Should probably fill this out

**Review by amzn-mike — APPROVED** [review]

**Pull request merged**